experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,479 @@
+    l = length(v)
+    x = 1
+    for i in 1:l
+        x = x*factorial(v[i])
+    end
+    return Int(factorial(n)/x)
+end
+
+##########################
+#Invariant Rings of Reductive groups
+##########################
+mutable struct InvariantRing
+    field::Field
+    poly_ring::MPolyDecRing #graded
+    
+    group::ReductiveGroup
+    representation::RepresentationReductiveGroup
+    
+    fundamental::Vector{MPolyDecRingElem}
+    primary::Vector{MPolyDecRingElem}
+    secondary::Vector{MPolyDecRingElem}
+    
+    reynolds_operator::Function
+    
+    NullConeIdeal::MPolyIdeal
+    
+    #Invariant ring of reductive group G (in representation R), no other input.
+    function InvariantRing(R::RepresentationReductiveGroup) #here G already contains information n and rep_mat
+        z = new()
+        n = ncols(R.rep_mat)
+        z.representation = R
+        z.group = R.group
+        G = z.group
+        z.field = G.field
+        z.poly_ring, __ = graded_polynomial_ring(G.field, "X" => 1:n)
+        z.reynolds_operator = reynolds_v_slm
+        return z
+    end
+    
+    #to compute invariant ring ring^G where G is the reductive group of R. 
+    function InvariantRing(R::RepresentationReductiveGroup, ring::MPolyRing)
+        n = ncols(R.rep_mat)
+        n == ngens(ring) || error("The given polynomial ring is not compatible.")
+        z = new()
+        z.representation = R
+        z.group = R.group
+        G = R.group
+        z.field = G.field
+        if is_graded(ring)
+            z.poly_ring = ring
+        else
+            z.poly_ring = grade(ring)[1]
+        end
+        z.reynolds_operator = reynolds_v_slm
+        return z
+    end
+end
+
+invariant_ring(R::RepresentationReductiveGroup) = InvariantRing(R)
+invariant_ring(R::RepresentationReductiveGroup, ring::MPolyRing) = InvariantRing(R, ring)
+null_cone_ideal(R::InvariantRing) = R.NullConeIdeal
+
+function fundamental_invariants(z::InvariantRing)
+    if isdefined(z, :fundamental)
+        return z.fundamental
+    else
+        R = z.representation
+        I, M = proj_of_image_ideal(R.group, R.rep_mat)
+        z.NullConeIdeal = ideal(generators(R.group, I, R.rep_mat))
+        z.fundamental = inv_generators(z.NullConeIdeal, R.group, z.poly_ring, M, I, z.reynolds_operator)
+        return z.fundamental
+    end
+end
+
+function Base.show(io::IO, R::InvariantRing) 
+    io = AbstractAlgebra.pretty(io)
+    println(io, "Invariant Ring of")
+    show(io, R.poly_ring)
+    print(io, AbstractAlgebra.Indent(),  " under group action of ", R.group.group[1], R.group.group[2])
+    print(io, AbstractAlgebra.Dedent())
+end
+
+#computing the graph Gamma from Derksens paper
+function image_ideal(G::ReductiveGroup, rep_mat::AbstractAlgebra.Generic.MatSpaceElem)
+    R = base_ring(rep_mat)
+    n = ncols(rep_mat)
+    m = G.group[2]
+    mixed_ring_xy, x, y, zz = PolynomialRing(G.field, "x"=>1:n, "y"=>1:n, "zz"=>1:m^2)
```

Comment:
The nested polynomial rings are not easier to use in this case, because I (think I) have to make new nested polynomials in the process. Here I don't make any new polynomial rings. If it's not too ugly, I'd prefer to keep it as it is.